src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -180,13 +180,22 @@ ShenandoahOldGeneration::ShenandoahOldGeneration(uint max_queues, size_t max_cap
     _promotable_humongous_regions(0),
     _promotable_regular_regions(0),
     _is_parseable(true),
+    _card_scan(nullptr),
     _state(WAITING_FOR_BOOTSTRAP),
     _growth_before_compaction(INITIAL_GROWTH_BEFORE_COMPACTION),
     _min_growth_before_compaction ((ShenandoahMinOldGenGrowthPercent * FRACTIONAL_DENOMINATOR) / 100)
 {
   _live_bytes_after_last_mark = ShenandoahHeap::heap()->capacity() * INITIAL_LIVE_FRACTION / FRACTIONAL_DENOMINATOR;
   // Always clear references for old generation
   ref_processor()->set_soft_reference_policy(true);
+
+  if (ShenandoahCardBarrier) {
+    // TODO: Old and young generations should only be instantiated for generational mode
```

Comment:
Yes, the old gen constructor should be able to assert that `ShenandoahCardBarrier` is on. The check is necessary because non-generational modes still instantiate the old generation.